docs/source/essentials/normalized-cache.mdx

```diff
@@ -0,0 +1,325 @@
+};
+
+//Build the ApolloClient
+ApolloClient apolloClient = ApolloClient.builder()
+  .serverUrl("https://...")
+  .normalizedCache(cacheFactory, resolver)
+  .build();
+```
+
+</MultiCodeBlock>
+
+For this resolver to work, every object in your graph needs to have a globally unique ID. If some of them don't have one, you can fallback to using the path as cache key by returning `CacheKey.NO_KEY`.
+
+## Using the cache with your queries
+
+You control how the cache is used with [`ResponseFetchers`](https://github.com/apollographql/apollo-android/blob/master/apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/ApolloResponseFetchers.java):
+
+<MultiCodeBlock>
+
+```kotlin
+// Get a response from the cache if possible. Else, get a response from the network
+// This is the default behavior
+val apolloCall = apolloClient().query(BookWithAuthorName()).responseFetcher(ApolloResponseFetchers.CACHE_FIRST)
+```
+
+```java
+// Get a response from the cache if possible. Else, get a response from the network
+// This is the default behavior
+ApolloCall apolloCall = apolloClient().query(new BookWithAuthorName()).responseFetcher(ApolloResponseFetchers.CACHE_FIRST)
+```
+
+</MultiCodeBlock>
+
+Other possibilities are CACHE_ONLY, NETWORK_ONLY, CACHE_AND_NETWORK_ONLY, NETWORK_FIST. See to the [`ResponseFetchers` class](https://github.com/apollographql/apollo-android/blob/master/apollo-runtime/src/main/java/com/apollographql/apollo/fetcher/ApolloResponseFetchers.java) for more details.
+
+## Reacting to changes in the cache
+
+One big advantage of using a normalized cache is that your UI can now react to changes in your cache data. If you want to be notified every time something changes in `book1`, you can use a `QueryWatcher`:
+
+<MultiCodeBlock>
+
+```kotlin
+  apolloClient.query(BookWithAuthorName()).watcher().toFlow().collect { response ->
+      // This will be called every time the book or author changes
+  }
+```
+
+```java
+  apolloClient.query(new BookWithAuthorName()).watcher().enqueueAndWatch(new ApolloCall.Callback<T>() {
+     @Override public void onResponse(@NotNull Response<T> response) {
+       // This will be called every time the book or author changes
+     }
+
+     @Override public void onFailure(@NotNull ApolloException e) {
+       // This will be called if an error happens
+     }
+  });
+```
+
+</MultiCodeBlock>
+
+## Interacting with the cache
+
+To manipulate the cache directly, `ApolloStore` exposes `read()` and `write()` methods:
+
+```kotlin
+  // Reading data from the store
+  val data = apolloClient.apolloStore.read(BookWithAuthorName()).execute()
+
+  // Create data to write
+  val data = BookWithAuthorName.Data(
+    id = "book1",
+    title = "Les guerriers du silence",
+    author = BookWithAuthorName.Author(
+      id = "author1",
+      name = "Pierre Bordage"
+    )
+  )
+  // Write to the store. All watchers will be notified
+  apolloClient.apolloStore.writeAndPublish(BookWithAuthorName(), data).execute()
+```
```

Comment:
Do you think we should add section like trouble shooting, what user should check first / pay attention to in case of cache miss or other issues?

Also we can mentioned here how to dump the cache content:
```
Map<KClass<?>, Map<String, Record>> dump = apolloClient.getApolloStore().normalizedCache().dump();
NormalizedCache.prettifyDump(dump)
```

Reply:
> Do you think we should add section like trouble shooting

That's definitely something that could be useful (See https://github.com/apollographql/apollo-android/issues/2397). Not sure what to recommend though, did you have something specific in mind?

> Also we can mentioned here how to dump the cache content

👍

Reply:
I think we should start with dump cache. Second we should mention to double check how cache key generated (make sure that there is stable ids). Third mention again about selection set should be the same or less. At least to start with these 3 points.

Maybe later we can expand more.